app.ts

```diff
@@ -235,28 +241,29 @@ async function attachToReport(reportId: string, fileUri: string) {
   await update(queryString);
 }
 
-app.get("/:id", async function (req, res) {
-  try {
-    const reportParts = await retrieveReportParts(req.params.id);
-    const reportContext = await retrieveContext(req.params.id);
-    const secretary = await retrieveReportSecretary(req.params.id);
-    if (!reportParts || !reportContext) {
+app.post("/generate", async function (req, res) {
+  
+  const reportIDs = req.body.reportIDs;
+  reportIDs.forEach(async reportID => {
+    try {  
+      const reportParts = await retrieveReportParts(reportID);
+      const reportContext = await retrieveContext(reportID);
+      const secretary = await retrieveReportSecretary(reportID);
+      if (!reportParts || !reportContext) {
+        res.status(500);
+        res.send("No report parts found.");
+        return;
+      }
+      const sanitizedParts = sanitizeReportParts(reportParts);
+      const fileMeta = await generatePdf(sanitizedParts, reportContext, secretary);
+      if (fileMeta) {
+        await attachToReport(reportID, fileMeta.uri);
+      }
+    } catch (e) {
       res.status(500);
-      res.send("No report parts found.");
-      return;
+      console.error(e);
+      res.send(e);
     }
-
-    const sanitizedParts = sanitizeReportParts(reportParts);
-    const fileMeta = await generatePdf(sanitizedParts, reportContext, secretary);
-    // await attachToReport(req.params.id, fileMeta.uri); // TODO: we do this in frontend now. Why?
-    if (fileMeta) {
-      res.send(fileMeta);
-      return;
-    }
-    throw new Error('Something went wrong while generating the pdf');
-  } catch (e) {
-    res.status(500);
-    console.error(e);
-    res.send(e);
-  }
+  }); 
+  res.sendStatus(200);
```

Comment:
**question**: what's the intention of using `async` inside a `forEach`?

The construct with `forEach(async () => {})` will result in the execution of the async functions but there's no actual promise to await and no way to ensure they all worked out. So in practice this code will boot up a bunch of async functions (one per report) and then immediately respond with a `200 OK` response. Is this intentional?

Although this will let you circumvent the issue with large payloads taking a lot of time to process, doing it like this gives the caller no way to check if the operation has succeeded or if any error occurred. In practice we'll always happily report success without actually knowing what's going on. It might be best to revisit this.

In any case, there's no point in doing anything with the response object (`res`) inside the async functions since by the time we reach errors we've already handled the request.